Document runner 2.0 feats (#623)
grammyjs/website: site/docs/plugins/runner.md

Instead, you can return _a list of constraint identifier strings_ that determine for every update individually what other computations it must wait for before processing can begin.

As an example, you could return both the chat identifier, and the user identifier of the message author.

```ts
bot.use(sequentialize((ctx) => {
  const chat = ctx.chat?.id.toString();
  const user = ctx.from?.id.toString();
  return [chat, user].filter((con) => con !== undefined);
}));
```

This would make sure that messages in the same chat are ordered correctly.
In addition, if Alice sends message in a group, and then sends a message to your bot in the private chat, then these two messages are ordered correctly.

In a sense, you can therefore specify a graph of dependencies between updates.
grammY runner will resolve all necessary constraints on the fly and block those updates as long as necessary to ensure correct message ordering.

The implementation of this is very efficient.
It needs constant memory (unless you specify infinite concurrency), and it needs (amortized) constant processing time per update.

## Graceful Shutdown

> In order for the bot to complete its work correctly you [should signal](../advanced/reliability.md#using-grammy-runner) it to stop when the process is about to be destroyed.

Reviewer: ```suggestion
In order for the bot to complete its work correctly, you [should signal it](../advanced/reliability.md#using-grammy-runner) to stop when the process is about to be destroyed.
```

> Note that you can wait for the runner to terminate by `await`ing the `task` in the [RunnerHandle](https://deno.land/x/grammy_runner/mod.ts?s=RunnerHandle) returned from `run`.

Reviewer: ```suggestion
Note that you can wait for the runner to terminate by `await`ing the `task` in the [`RunnerHandle`](https://deno.land/x/grammy_runner/mod.ts?s=RunnerHandle) returned from `run`.
```

```ts
const handle = run(bot);

handle.task.then(() => {
  console.log("Bot done processing!");
});
```

## Advanced Options

> grammY runner consists of three things: a source, a runner, and a sink.

Reviewer: Maybe it’ll be cooler like this since it will match the order of the explanation following it?

```suggestion
grammY runner consists of three things: a source, a sink, and a sink.
```

Author: It will have to be this:

```suggestion
grammY runner consists of three things: a source, a sink, and a runner.
```

The source pulls in updates, the sink consumes updates, and the runner configures and connects the two.

> > An in-depth description on how the runner workds internally can be found [down here](#how-it-works-behind-the-scenes).

Reviewer: Workd?

Author: ```suggestion
> An in-depth description on how the runner works internally can be found [down here](#how-it-works-behind-the-scenes).
```

Reviewer: Good.

Each of these three parts can be configured through various options.
This can reduce the network traffic, let you specify allowed updates, and more.

Each part of the runner accepts its configuration through a dedicated options object.

```ts
run(bot, {
  source: {},
  runner: {},
  sink: {},
});
```

You should check out the `RunOptions` in the [API reference](https://deno.land/x/grammy_runner/mod.ts?s=RunOptions) to see which options are available.

> For example, you will find there that `allowed_updates` can be enabled using the following code snippet.

Reviewer: ```suggestion
For example, you will there find out that `allowed_updates` can be enabled using the following code snippet.
```

```ts
run(bot, { runner: { fetch: { allowed_updates: [] } } });
```

## Multithreading

JavaScript is single-threaded.

> This is amazing because [concurrency is hard](../advanced/scaling.md#concurrency-is-hard) so if there is only a single thread, then this removes a lot of headache.

Reviewer: ```suggestion
This is amazing because [concurrency is hard](../advanced/scaling.md#concurrency-is-hard), meaning that if there is only a single thread, a lot of headache is automatically removed.
```

Author: Is automatically removed? Sounds like some technical automation

Reviewer: ```suggestion
This is amazing because [concurrency is hard](../advanced/scaling.md#concurrency-is-hard), meaning that if there is only a single thread, a lot of headache is naturally removed.
```

> However, if your bot has extremely high load (we are talking about 1000 updates per second and up), then doing everything on a single core might not be enough anymore.

Reviewer: ```suggestion
However, if your bot has an extremely high load (we are talking about 1000 updates per second and up), then doing everything on a single core might not be enough anymore.
```

Basically, a single core will start struggling with the JSON processing of all the messages your bot has to handle.

### Bot Workers for Update Handling

> There is a simple way out: bot workers!

Reviewer: ```suggestion
There is a simple way out: Bot workers!
```

Author: Isn't the first word capitalised iff the part following the colon is a sentence?

Reviewer: Right.

grammY runner lets you create several workers which can process your updates in parallel on actually different cores (using different event loops and with separate memory).

On Node.js, grammY runner uses [Worker Threads](https://nodejs.org/api/worker_threads.html).
On Deno, grammY runner uses [Web Workers](https://deno.land/manual/runtime/workers).

> Concuptually, grammY runner provides you with a class called `BotWorker` which can handle updates.

Author: ```suggestion
Conceptually, grammY runner provides you with a class called `BotWorker` which can handle updates.
```

Reviewer 2: Outdate

Reviewer 2: [9efa9dc](https://github.com/grammyjs/website/pull/623/commits/9efa9dcf4fa0093cc1fce192d09252354e55994f)

It is equivalent to the regular class `Bot` (in fact, it even `extends Bot`).
The main difference between `BotWorker` and `Bot` is that `BotWorker` cannot fetch updates.
Instead, it has to receive them from a regular `Bot` that controls its workers.

```asciiart:no-line-numbers
1. fetch updates                                 Bot
                                              __// \\__
                                           __/  /   \  \__
2. send updates to workers              __/    /     \    \__
                                     __/      /       \      \__
                                    /        /         \        \
3. process updates          BotWorker   BotWorker   BotWorker   BotWorker
```

grammY runner provides you with middleware that can send updates to bot workers.
The bot workers can then receive this update and handle it.
This way, the central bot only has to concern itself with pulling in and distributing updates among the bot workers it orchestrates.

> The actual update handling (filtering messages, sending replies, etc) is performed by the bot workers.

Reviewer: ```suggestion
The actual update handling (filtering messages, sending replies, etc.) is performed by the bot workers.
```

Author: TIL

Let's now see how this can be used.

### Using Bot Workers

> Examples of this can be found in the [grammY runner repository](https://github.com/grammyjs/runner/tree/main/examples).

We will start out by creating the central bot instance that fetches updates and distributes them among workers.

> Let's create a file called `bot.ts` with the following code.

Reviewer: ```suggestion
Let's start by creating a file called `bot.ts` with the following content.
```

<CodeGroup>
  <CodeGroupItem title="TypeScript" active>

```ts
// bot.ts
import { Bot } from "grammy";
import { distribute, run } from "@grammyjs/runner";

// Create the bot.
const bot = new Bot("");

// Optionally, sequentialize updates here.
// bot.use(sequentialize(...))

// Distribute the updates among bot workers.
bot.use(distribute(__dirname + "/worker"));

> // Run the concurrently with multi-threading.

Reviewer: Run the concurrently?

Author: ```suggestion
// Run the bot concurrently with multi-threading.
```

```ts
// bot.ts
const { Bot } = require("grammy");
const { distribute, run } = require("@grammyjs/runner");

// Create the bot.
const bot = new Bot("");

// Optionally, sequentialize updates here.
// bot.use(sequentialize(...))

// Distribute the updates among bot workers.
bot.use(distribute(__dirname + "/worker"));

// Run the concurrently with multi-threading.
run(bot);
```

</CodeGroupItem>
 <CodeGroupItem title="Deno">

```ts
// bot.ts
import { Bot } from "https://deno.land/x/grammy/mod.ts";
import { distribute, run } from "https://deno.land/x/grammy_runner_/mod.ts";

// Create the bot.
const bot = new Bot("");

// Optionally, sequentialize updates here.
// bot.use(sequentialize(...))

// Distribute the updates among bot workers.
bot.use(distribute(new URL("./worker.ts", import.meta.url)));

// Run the concurrently with multi-threading.
run(bot);
```

</CodeGroupItem>
</CodeGroup>

Next to `bot.ts`, we create a second file called `worker.ts` (as specified on line 12 in the code above).
This will contain the actual bot logic.

<CodeGroup>
  <CodeGroupItem title="TypeScript" active>

```ts
// worker.ts
import { BotWorker } from "@grammyjs/runner";

// Create a new bot worker.

> const bot = new BotWorker(""); // <-- pass your bot token here again

Author: I'm not so sure about this one. We want to emphasize that the token must be passed twice. Perhaps it's good to be very different here?

> const bot = new BotWorker(""); // <-- pass your bot token here again

Reviewer: Why not just `bot.token`?

> const bot = new BotWorker(""); // <-- pass your bot token here again

Author: It's in a different file and you really, really don't want to add the dependency because then you'd have two files that import each other.

If anything, we could consider injecting it with the first update, but this would be less explicit, harder to understand, and more complex to implement.

> const bot = new BotWorker(""); // <-- pass your bot token here again

Reviewer: Oh then I was fucking wrong. We just don't need the `again`.

> const bot = new BotWorker(""); // <-- pass your bot token here again

Reviewer: We simply won’t.

> const bot = new BotWorker(""); // <-- pass your bot token here again

Author: Can we add a note below the code snippet which points this out?

> const bot = new BotWorker(""); // <-- pass your bot token here again

Author: Nvm, if anything, I'll do this in subsequent PQs. This is too late here

> const bot = new BotWorker(""); // <-- pass your bot token here again

Reviewer 3: Agree with @KnorpelSenf. I prefer to stick with "pass your bot token here again" to make it clear the token needs to be inserted again. Since it's already explained in the previous code, it's unlikely to confuse the reader.

// Add message handling logic.
bot.on("message", (ctx) => ctx.reply("yay!"));
```

</CodeGroupItem>
 <CodeGroupItem title="JavaScript">

```ts
// worker.ts
const { BotWorker } = require("@grammyjs/runner");

// Create a new bot worker.
const bot = new BotWorker(""); // <-- pass your bot token here again

// Add message handling logic.
bot.on("message", (ctx) => ctx.reply("yay!"));
```

</CodeGroupItem>
 <CodeGroupItem title="Deno">

```ts
// worker.ts
import { BotWorker } from "https://deno.land/x/grammy_runner/mod.ts";

// Create a new bot worker.

> const bot = new BotWorker(""); // <-- pass your bot token here again

Author: ditto

(This handle is also returned by `run`.)
Check out the [API reference](https://deno.land/x/grammy_runner/mod.ts?s=RunnerHandle) of the `RunnerHandle`.

### The `run` Function

The `run` function does a few things to help you use the above structure with ease.

1. It creates an update supplier from your bot.
2. It creates a [source](#source) from the update supplier.
3. It creates an update consumer from your bot.
4. It creates a [sink](#sink) from the update consumer.

> 5. It creates a runner from the source and the sink.

Reviewer 2: Create link like: [runner](#runner)

> 5. It creates a runner from the source and the sink.

Reviewer 3: ```suggestion
5. It creates a [runner](#runner) from the source and the sink.
```
